pkg/reconciler/finalizer/finalizer.go

```diff
@@ -0,0 +1,605 @@
+		}
+		// Copy updates back into the original unstructured object
+		uObj, err := kinds.ToUnstructured(rObj, core.Scheme)
+		if err != nil {
+			return err
+		}
+		uObj.DeepCopyInto(obj)
+		return nil
+	})
+	if err != nil {
+		return updatedObj, updated, errors.Wrapf(err, "failed to remove ReconcilerFinalizing condition")
+	}
+	if updated {
+		klog.Info("ReconcilerFinalizing condition removal successful")
+	} else {
+		klog.Warning("ReconcilerFinalizing condition removal skipped: already removed")
+	}
+	return updatedObj, updated, nil
+}
+
+// errorHandler logs errors from the Informer's ListAndWatch.
+// Based on DefaultWatchErrorHandler from k8s.io/client-go@v0.23.2/tools/cache/reflector.go
+func (c *Controller) errorHandler(r *cache.Reflector, err error) {
+	switch {
+	// Stop channel closed
+	case err == io.EOF:
+		klog.Infof("ListAndWatch error (termination expected): %v", err)
+
+	// Watch connection closed
+	case err == io.ErrUnexpectedEOF:
+		klog.V(1).Infof("ListAndWatch error (retry expected): %v", err)
+
+	// Context done
+	case errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded):
+		klog.Infof("ListAndWatch error (termination expected): %v", err)
+
+	// resourceVersion too old
+	case apierrors.IsResourceExpired(err):
+		klog.V(1).Infof("ListAndWatch error (retry expected): %v", err)
+
+	// Resource unregistered (DEPRECATED, see NotFound)
+	case apierrors.IsGone(err):
+		klog.V(1).Infof("ListAndWatch error (retry expected): %v", err)
+
+	// Resource not registered
+	case apierrors.IsNotFound(err):
+		klog.V(1).Infof("ListAndWatch error (retry expected): %v", err)
+
+	// Insufficient permissions
+	case apierrors.IsForbidden(err):
+		klog.V(1).Infof("ListAndWatch error (retry expected): %v", err)
+
+	// Unexpected error
+	default:
+		klog.Warningf("ListAndWatch error (retry expected): %v", err)
+	}
+}
+
+type eventHandler struct {
+	controller *Controller
+	context    context.Context
+
+	// cancel will be called by the finalize() method to stop the parser and remediator.
+	cancel context.CancelFunc
+
+	// continueCh will be closed by the caller when the parser and remediator
+	// have fully stopped. This unblocks finalize() to destroy managed resource objects.
+	continueCh <-chan struct{}
+
+	// finalizing is used as an atomic boolean.
+	// 0 means the finalize() method is not running.
+	// 1 means the finalize() method is running.
+	// When finalize() is running, all concurrent calls to finalize() will return immediately.
+	finalizing uint32
+}
+
+// OnAdd adds/removes the finalizer as-needed, depending on the annotation,
+// and handles triggering the finalizer if the deletionTimestamp and finalizer
+// are set.
+func (eh *eventHandler) OnAdd(newIObj interface{}) {
+	newObj, ok := newIObj.(*unstructured.Unstructured)
+	if !ok {
+		panic(fmt.Sprintf("OnAdd received unexpected object type %T", newIObj))
+	}
+	klog.V(1).Infof("OnAdd %s", log.AsJSON(newObj))
+
+	obj := newObj.DeepCopy()
+	ctx := eh.context
```

Comment:
Do we need the extra `ctx` parameter for the `finalize`/`reconcilerFinalizer` methods? Looks like they have access to it through the pointer receiver

Reply:
Debatable. I don't really like passing the context in as a struct field. It's bad for readability/testability, but it was required to be able to satisfy the event handler interface. I think to clean it up I might just move those functions to the Controller, so they're not cluttering up the eventHandler, which is more of a helper subclass.

Reply:
I think I'm gonna refactor the `Controler` and `eventHandler` and extract a `Finalizer` class to hold most of the action methods. That way the Controller & eventHandler can be tested independently of the Finalizer.

Reply:
Done.

The event handling is now done by the controller-manager, which converts them into Controller.Reconciler calls, with an execution and retry queue. So the context passed to Manager.Start gets propagated to Controller.Reconciler.